commons-macros/src/main/scala/com/avsystem/commons/macros/serialization/GenCodecMacros.scala

```diff
@@ -258,30 +253,85 @@ class GenCodecMacros(ctx: blackbox.Context) extends CodecMacroCommons(ctx) with
       def generatedWrite(sym: Symbol) =
         q"writeField(${nameBySym(sym)}, output, ${mkParamLessCall(q"value", sym)}, ${depNames(sym)})"
 
+      def readField(param: ApplyParam) = {
+        val fieldName = nameBySym(param.sym)
+        val defaultValueTree = param.defaultValue.orElse(q"fieldMissing($fieldName)")
+        if (!isScalaJs) q"${optName(param)}.getOrElse($defaultValueTree)"
+        else {
+          if (param.valueType != weakTypeOf[Unit])
+            q"""
+              if (!_root_.scala.scalajs.js.isUndefined(elements.selectDynamic($fieldName))) {
+                elements.selectDynamic($fieldName).asInstanceOf[${param.valueType}]
+              } else ($defaultValueTree).asInstanceOf[${param.valueType}]
+            """
+          else q"()"
+        }
+      }
+
+      def generateReadObject =
+        if (!isScalaJs) {
+          q"""
+            def readObject(input: $SerializationPkg.ObjectInput, ..$oooFieldsParam): $tpe = {
+              ..${params.map(varDeclaration)}
+              while(input.hasNext) {
+                val fi = input.nextField()
+                fi.fieldName match {
+                  case ..${params.map(fieldReadCase)}
+                  case _ => fi.skip()
+                }
+              }
+              ${applier(params.map(p => p.asArgument(readField(p))))}
+            }
+          """
+        } else {
+          q"""
+            def readObject(input: $SerializationPkg.ObjectInput, ..$oooFieldsParam): $tpe = {
+              val elements: _root_.scala.scalajs.js.Dynamic = _root_.scala.scalajs.js.Object().asInstanceOf[_root_.scala.scalajs.js.Dynamic]
+              while(input.hasNext) {
+                val fi = input.nextField()
+                fi.fieldName match {
+                  case ..${params.map { p =>
+                    val fieldName = nameBySym(p.sym)
+                    cq"""
+                      $fieldName =>
+                        elements.updateDynamic($fieldName)(
+                          readField(fi, ${depNames(p.sym)}).asInstanceOf[_root_.scala.scalajs.js.Any]
+                        )
+                    """
+                  }}
+                  case _ => fi.skip()
+                }
+              }
+              ..${
+                params.collect { case p if oooFields.contains(nameBySym(p.sym)) =>
+                  val name = nameBySym(p.sym)
+                  val varName = optName(p)
+                  q"""
+                    if (oooFields.$varName.isDefined)
+                      elements.updateDynamic($name)(oooFields.$varName.get.asInstanceOf[_root_.scala.scalajs.js.Any])
+                  """
+                }
+              }
+              ${applier(params.map(p => p.asArgument(readField(p))))}
+            }
+          """
+        }
+
       q"""
-        new $baseClass with $GenCodecObj.ErrorReportingCodec[$tpe] {
+        new $baseClass(
+          ${tpe.toString},
```

Comment:
To avoid unnecessary fields, I would prefer if these were not passed as arguments but as protected method implementations. My intuition is that this should not be a problem with regard to generated code size. But of course correct me if I'm wrong.

Reply:
Method implementation generates code for each created anonymous class.